File: MuscleAssert/ObjC/Differ/MUSStringDiffer.m
//
//  MuscleAssert.swift
//  MuscleAssert
//
//  Created by akuraru on 2016/12/17.
//
//

#import "MUSStringDiffer.h"
#import "MUSDifference.h"

NS_ASSUME_NONNULL_BEGIN

@implementation MUSStringDiffer

- (Class)matchClass {
    return [NSString class];
}

- (NSArray<MUSDifference *> *)diff:(id)left right:(id)right path:(nullable NSString *)path delegatge:(id<MUSDeepDiffProtocol>)delegate {
    return [self stringDiff:right left:left path:path];
}

- (NSArray<MUSDifference *> *)stringDiff:(NSString *)right left:(NSString *)left path:(NSString *_Nullable)path {
    NSArray *diff = [self lcsDiff:right right:left];
    NSInteger length = [diff count];
    
    NSMutableArray *result = [NSMutableArray array];
    for (NSInteger index = 0; index < length; index++) {
        [result addObject:[[MUSDifference alloc] initWithPath:[self pathByAppendingPath:path index:[diff[index][0] integerValue]]left:diff[index][2] right:diff[index][1] ]];
    }
    return result;
}

- (NSArray *)lcsDiff:(NSString *)left right:(NSString *)right {
    NSString *lcs = [self longestCommonSubsequence:left right:right];
    NSUInteger l1 = [left length];
    NSUInteger l2 = [right length];
    NSUInteger lc = [lcs length];
    NSUInteger idx1 = 0;
    NSUInteger idx2 = 0;
    NSUInteger idxc = 0;
    NSMutableString *s1 = [[NSMutableString alloc] initWithCapacity:l1];
    NSMutableString *s2 = [[NSMutableString alloc] initWithCapacity:l2];
    NSMutableArray *res = [NSMutableArray arrayWithCapacity:10];
    for (;;) {
        if (idxc >= lc) break;
        unichar c1 = [left characterAtIndex:idx1];
        unichar c2 = [right characterAtIndex:idx2];
        unichar cc = [lcs characterAtIndex:idxc];
        if ((c1 == cc) && (c2 == cc)) {
            if ([s1 length] || [s2 length]) {
                NSArray *e = @[@(idxc), s1, s2];
                [res addObject:e];
                s1 = [[NSMutableString alloc] initWithCapacity:l1];
                s2 = [[NSMutableString alloc] initWithCapacity:l1];
            }
            idx1++;
            idx2++;
            idxc++;
            continue;
        }
        if (c1 != cc) {
            [s1 appendString:[NSString stringWithCharacters:&c1 length:1]];
            idx1++;
        }
        if (c2 != cc) {
            [s2 appendString:[NSString stringWithCharacters:&c2 length:1]];
            idx2++;
        }
    }
    if (idx1 < l1) {
        [s1 appendString:[left substringFromIndex:idx1]];
    }
    if (idx2 < l2) {
        [s2 appendString:[right substringFromIndex:idx2]];
    }
    if ([s1 length] || [s2 length]) {
        NSArray *e = @[@(idxc), s1, s2];
        [res addObject:e];
    }
    return res;
}

- (NSString *)longestCommonSubsequence:(NSString *)left right:(NSString *)right {
    const NSUInteger leftLength = left.length;
    const NSUInteger rightLength = right.length;
    
    unsigned int **lengths = malloc((leftLength + 1) * sizeof(unsigned int *));
    
    for (unsigned int i = 0; i <= leftLength; ++i) {
        lengths[i] = malloc((rightLength + 1) * sizeof(unsigned int));
        
        for (unsigned int j = 0; j <= rightLength; ++j) {
            lengths[i][j] = 0;
        }
    }
    
    
    for (unsigned int i = 0; i < leftLength; ++i) {
        for (unsigned int j = 0; j < rightLength; ++j) {
            if ([left characterAtIndex:i] == [right characterAtIndex:j]) {
                lengths[i + 1][j + 1] = lengths[i][j] + 1;
            } else {
                lengths[i + 1][j + 1] = MAX(lengths[i + 1][j], lengths[i][j + 1]);
            }
        }
    }
    
    NSMutableString *lcs = [NSMutableString string];
    NSUInteger x = leftLength;
    NSUInteger y = rightLength;
    
    while (x != 0 && y != 0) {
        if (lengths[x][y] == lengths[x - 1][y]) {
            --x;
        } else if (lengths[x][y] == lengths[x][y - 1]) {
            --y;
        } else {
            [lcs insertString:[NSString stringWithFormat:@"%C", [left characterAtIndex:x - 1]] atIndex:0];
            --x;
            --y;
        }
    }
    
    for (unsigned int i = 0; i <= leftLength; ++i) {
        free(lengths[i]);
    }
    
    free(lengths);
    
    return lcs;
}

@end

NS_ASSUME_NONNULL_END

File: MuscleAssert/Swift/Differ/MUSStringDiffer.swift
//
//  MuscleAssert.swift
//  MuscleAssert
//
//  Created by akuraru on 2016/12/17.
//
//

extension String.Index: Hashable {
    public var hashValue: Int {
        return self.encodedOffset
    }
}

class MUSStringDiffer: MUSCustomClassDiffer {
    typealias MatchType = String
    
    func diff(left: Any, right: Any, path: String?, delegatge: MUSDeepDiffProtocol) -> [MUSDifference] {
        guard let left = left as? String, let right = right as? String else {
            fatalError("none reach")
        }
        let diff = lcsDiff(left: left, right: right)
    
        return diff.map { d in
            MUSDifference(path: pathByAppendingPath(path:path, index:d.index), left:d.left, right:d.right)
        }
    }
    func lcsDiff(left: String, right: String) -> [(index: Int, left: String, right: String)] {
        let lcs = longestCommonSubsequence(left: left, right: right)
        let l1 = left.endIndex
        let l2 = right.endIndex
        let lc = lcs.endIndex
        var idx1 = left.startIndex;
        var idx2 = right.startIndex;
        var idxc = lcs.startIndex;
        var s1 = String()
        var s2 = String()
        var res = [(index: Int, left: String, right: String)]()
        while true {
            if idxc >= lc { break }
            let c1 = left[idx1]
            let c2 = right[idx2]
            let cc = lcs[idxc]
            if ((c1 == cc) && (c2 == cc)) {
                if !s1.isEmpty || !s2.isEmpty {
                    res.append((idxc.encodedOffset, s1, s2))
                    s1 = String()
                    s2 = String()
                }
                idx1 = left.index(after: idx1)
                idx2 = right.index(after: idx2)
                idxc = lcs.index(after: idxc)
                continue;
            }
            if (c1 != cc) {
                s1.append(c1)
                idx1 = left.index(after: idx1)
            }
            if (c2 != cc) {
                s2.append(c2)
                idx2 = right.index(after: idx2)
            }
        }
        if (idx1 < l1) {
            s1.append(String(left[idx1..<left.endIndex]))
        }
        if (idx2 < l2) {
            s2.append(String(right[idx2..<right.endIndex]))
        }
        if !s1.isEmpty || !s2.isEmpty {
            res.append((idxc.encodedOffset, s1, s2))
        }
        return res;
    }
    func longestCommonSubsequence(left : String, right: String) -> String {
        var lengths = [String.Index: [String.Index: Int]]()
        
        lengths[left.startIndex] = [String.Index: Int]()
        for leftIndex in left.indices {
            lengths[left.index(after: leftIndex)] = [String.Index: Int]()
            for rightIndex in right.indices {
                if left[leftIndex] == right[rightIndex] {
                    lengths[left.index(after: leftIndex)]![right.index(after: rightIndex)] = (lengths[leftIndex]?[rightIndex] ?? 0) + 1
                } else {
                    lengths[left.index(after: leftIndex)]![right.index(after: rightIndex)] = max(lengths[left.index(after: leftIndex)]?[rightIndex] ?? 0, lengths[leftIndex]?[right.index(after: rightIndex)] ?? 0)
                }
            }
        }
        
        var lcs = [Character]()
        var x = left.endIndex
        var y = right.endIndex
        
        while (x != left.startIndex && y != right.startIndex) {
            if (lengths[x]![y] ?? 0 == lengths[left.index(before: x)]![y] ?? 0) {
                x = left.index(before: x)
            } else if (lengths[x]![y] ?? 0 == lengths[x]![right.index(before: y)] ?? 0) {
                y = right.index(before: y)
            } else {
                x = left.index(before: x)
                y = right.index(before: y)
                lcs.append(left[x])
            }
        }
        
        return String(lcs.reversed())
    }
}

